crates/ruma-identifiers/src/mxc_uri.rs

```diff
@@ -14,39 +14,52 @@ use crate::ServerName;
 #[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct MxcUri {
     full_uri: Box<str>,
-    slash_idx: Option<NonZeroU8>,
 }
 
 impl MxcUri {
     /// If this is a valid MXC URI, returns the media ID.
     pub fn media_id(&self) -> Option<&str> {
-        self.parts().map(|(_, s)| s)
+        self.parts().ok().map(|(_, s)| s)
     }
 
     /// If this is a valid MXC URI, returns the server name.
     pub fn server_name(&self) -> Option<&ServerName> {
-        self.parts().map(|(s, _)| s)
+        self.parts().ok().map(|(s, _)| s)
     }
 
-    /// If this is a valid MXC URI, returns a `(server_name, media_id)` tuple.
-    pub fn parts(&self) -> Option<(&ServerName, &str)> {
-        self.slash_idx.map(|idx| {
+    /// If this is a valid MXC URI, returns a `(server_name, media_id)` tuple, else it returns the
+    /// error.
+    pub fn parts(&self) -> Result<(&ServerName, &str), MxcUriError> {
+        self.extract().map(|idx| {
             (
                 <&ServerName>::try_from(&self.full_uri[6..idx.get() as usize]).unwrap(),
                 &self.full_uri[idx.get() as usize + 1..],
             )
         })
     }
 
-    /// Returns if this is a spec-compliant MXC URI.
+    /// Validates the URI and returns an error if it failed.
+    pub fn validate(&self) -> Result<(), MxcUriError> {
+        self.extract().map(|_| ())
+    }
+
+    /// Convenience method for `.validate().is_ok()`
+    #[inline(always)]
     pub fn is_valid(&self) -> bool {
-        self.slash_idx.is_some()
+        self.validate().is_ok()
     }
 
     /// Create a string slice from this MXC URI.
+    #[inline(always)]
     pub fn as_str(&self) -> &str {
         &self.full_uri
     }
+
+    // convenience method for calling validate(self)
+    #[inline(always)]
+    fn extract(&self) -> Result<NonZeroU8, MxcUriError> {
```

Comment:
I think this name is a bit too vague. Maybe `extract_slash_idx`?

Reply:
its internal, but will do 👍